### ISSUE TITLE
ID every item on pickup?

### ISSUE THREAD

**Issue by dakky221** [kolton/d2bot-with-kolbot#2824]
This was already asked and discussed but so far I haven't found a solution, some say that town.js or pickit.js should be edited, but to be honest I have no idea what and where.

So the idea is to pick all unique items and identify them before stashing/automuling.

Putting something like this in nip file...

[quality] == unique # [fcr] == 99
[quality] == unique

...doesn't work. That is in "alluniques.nip" and _only that file is pushed_ in character config. It's loaded and while it picks every unique item, it doesn't identify them. The idea was to use the first line as something impossible as any item has to be identified to be checked for FCR and when the pickit sees it's not fcr of 99, it should move to the next line where it now keeps the identified item. But that doesn't work, pickit just skips the first line and keeps the unique without identifying.

So... what to put in pickit.js so it either fieldID the unique item (and logs it as identified) OR doesn't log it as kept until it identifies it at NPC? The end result should be the same - kept identified unique item.

**Comment by TrazhCant** [comment]
[quality] == unique # [fcr] != 99

delete the second line

**Comment by dakky221** [comment]
Nope, doesn't work. Still picks the unique and doesn't identify it.

![image](https://user-images.githubusercontent.com/22996638/98431547-99b75e00-20b6-11eb-9321-1eedf7311e3c.png)

Only that what you said is in nip file. Edit: it kept it and stashed unidentified too. I guess the bot knows that mace can't have fcr 99 so it omits that entry. Don't see other explanation.

**Comment by TrazhCant** [comment]
Possible.  try it with another attribute like this

[quality] == unique # [strength] != 1

True, maces do not have an affix for fcr (except runeword like HOTO).  Nothing unique has a strength of +1 but every item slot has eligibility for an affix that can have strength on magic items, so this should prove/disprove this theory.

**Comment by TrazhCant** [comment]
If only having this line doesn't work:

`[quality] == unique # [strength] != 1`

You might need to put something for each slot maybe like this instead:

`[type] == armor && [quality] == unique # [strength] != 1`
`[type] == anyarmor && [quality] == unique # [strength] != 1`
`[type] == weapon && [quality] == unique # [strength] != 1`
`[type] == helm && [quality] == unique # [strength] != 1`
`[type] == shield && [quality] == unique # [strength] != 1`
`[type] == anyshield && [quality] == unique # [strength] != 1`
`[type] == secondhand && [quality] == unique # [strength] != 1`
`[type] == boots && [quality] == unique # [strength] != 1`
`[type] == gloves && [quality] == unique # [strength] != 1`
`[type] == belt && [quality] == unique # [strength] != 1`
`[type] == ring && [quality] == unique # [strength] != 1`
`[type] == amulet && [quality] == unique # [strength] != 1`
`[type] == jewel && [quality] == unique # [strength] != 1`
`[type] == classspecific && [quality] == unique # [strength] != 1`
`[name] == smallcharm && [quality] == unique # [strength] != 1 # [maxquantity] == 1`
`[name] == largecharm && [quality] == unique # [strength] != 1 # [maxquantity] == 1`
`[name] == grandcharm && [quality] == unique # [strength] != 1 # [maxquantity] == 1`


The charms lines are to make sure you don't try to pick another gheeds/anni/torch if you already have one.  If your bot isn't farming those, then they're unnecessary.  Also, some of these entries might be redundant, but these item types should cover every possible unique in the game.  Some redundancy might be like the armor/anyarmor or shield/anyshield entries but I'm not sure.  Also, I put secondhand to cover Necro heads, but that could also be covered by the classspecific entry.  Without testing, I'm not sure which might be unnecessary for sure, so I just included all of them for this suggestion.  Do post a reply if you try this or figure out the answer.  I'm interested to know how to possibly do this, too. :)

**Comment by dakky221** [comment]
Still doesn't work... with all lines as you wrote, now it happened this:

![image](https://user-images.githubusercontent.com/22996638/98441452-afe90c80-20fe-11eb-839f-3a6a32f7b08b.png)

It won't identify the item no matter the criteria in the nip file. Check other posts on this subject, someone said it won't work without editing pickit.js to force identification on pickup or on NPC visit. Edit: we're blind, I don't know if [strength] is also required strength, but that can also be seen without ID ^^. It's not 1 is it? :)

**Comment by TrazhCant** [comment]
Hmm, that sucks.  And no, [strength] does not refer to required strength.  You could try replacing [strength] with [maxhp] but I doubt that will change anything.

Has any of this given you any kind of change, or have all the results been the exact same?

**Comment by dakky221** [comment]
All is the same, always. It just picks the unidentified item and stashes it as unidentified (and later automules unidentified). This was already asked [here](https://github.com/kolton/d2bot-with-kolbot/issues/2545), and never was solved. It obviously involves changing the pickit.js.

**Comment by TrazhCant** [comment]
In your character config, do you have these lines?:
(Approximately lines 320-322)
`// Additional item info log settings. All info goes to \logs\ItemLog.txt`
`Config.ItemInfo = false; // Log stashed, skipped (due to no space) or sold items.`
`Config.ItemInfoQuality = []; // The quality of sold items to log. See NTItemAlias.dbl for values. Example: Config.ItemInfoQuality = [6, 7, 8];`
(Approximately line 340)
`Config.SkipLogging = []; // Custom log skip list. Set as three digit item code or classid. Example: ["tes", "ceh", 656, 657] will ignore logging of essences.`

In pickit.js, do you have this line? It should be near the top like line 10:
`ignoreLog: [4, 5, 6, 22, 41, 76, 77, 78, 79, 80, 81], // Ignored item types for item logging`

If all of this checks out, then I'm stumped.

**Comment by dakky221** [comment]
Yep, its exactly the same. In char config and in pickit.js as you wrote.

It is not a bug, but rather a feature, because some items are more worth unidentified than identified. Forcing them to identify would break that idea of selling unids and that's why pickit works the way it works.

After all, you can all that test by yourself, I'm pretty sure you'll get the same end results.

**Comment by DarkHorseDre** [comment]
ok so I had a quick look.

There seems to be the sequence of:
nip (rules to pick) --> pickit (canfit etc) --> town (when to stash etc) --> storage (called from town to `moveto` item to stash)

As Dakky is not afraid to help himself, I will leave to fine tuning to him :) 

in `Town.stash` you can make the following changes to the function to **test**:
```js
if (result) {
	var tome = me.findItem(519, 0, 3); //dhd start - assume there needs to be ID tome
	if (tome || tome.getStat(70) > 1) { //assume you need at least 1 scroll in tome - I'd say increase this and test what happens if no scroll (go to shop to refill, quit or?)
		this.identifyItem(items[i], tome);  //id before stashing
        }//dhd end
	Misc.itemLogger("Stashed", items[i]);
	Storage.Stash.MoveTo(items[i]);
```

you also need to test what happens during fieldid, cainid, makeroom town trips, and if it keeps items that are supposed to be unid (if not you need to copy/call that code to override this)

here is the pickit line I used to test:
```js
([type] == helm || [type] == armor || [type] == gloves || [type] == boots || [type] == ring || [type] == belt || [type] == amulet || [type] == shield) && [quality] >= magic # [strength] >= 0
```

oh, and I placed my code before the logging parts to try to log the unid items, but ofc the list was not updated so they still show unid.

**Comment by dakky221** [comment]
This worked. It still shows in manager as unid, but ID's when putting in stash. Will test to see what happens when it runs out of scrolls or when the stash is full. I'll get back when it runs for a while :)  ... edit... actually it identifies it as soon as it gets to town, before putting in stash. So maybe if the stash is full it will be identified and stay in inventory. Good job so far! :)

edit: if the stash is full, doesn't even check if there is space (guess it knows it doesn't have space) so it doesn't identifies the item.

**Comment by DarkHorseDre** [comment]
yeah, check my post detail - it id's as part of moving to stash.. you can just move it to wherever it is checked (before dropping, selling etc) or simply id as it drops for muling.. depends on what you want but its all open - you have a place to start from now.

**Comment by dakky221** [comment]
I've tried pretty much everything, even copied whole FieldID function to the pickit.js, called your function there, it just won't identify on pickup and in town does nothing if item is stuck in inventory (full stash). Doesn't identify either on fillTome at NPC, wherever I put it.

**Comment by dakky221** [comment]
Changed line 554 to 560 in town.js into this

`				// Force ID for unid items matching autoEquip criteria
				if (result.result === 1 && !item.getFlag(0x10) && Item.hasTier(item)) {
					var tome = me.findItem(519, 0, 3); //dhd start - assume there needs to be ID tome
					if (tome || tome.getStat(70) > 1) { //assume you need at least 1 scroll in tome - I'd say increase this and test what happens if no scroll (go to shop to refill, quit or?)
						this.identifyItem(item, tome);  //id before stashing
					}//dhd end
					//result.result = -1;`

and it identified the item in town while the stash is full ! Needs further testing !
